### PR TITLE
Issue 45

### PR DESCRIPTION
**Pull request by timerplayer**
A pull request that fixes an issue #45
### Description
This pull request allow to use few sessions to Jira server in same time.
For ex.:
```
Get-Issue -Key "AA-0000" -Session $sessionA
Get-Issue -Key "AA-0000" -Session $sessionB
Get-Issue -Key "AA-0000" -Session "SessionNameA"
Get-Issue -Key "AA-0000" -Session "SessionNameB"
```

### Motivation and Context
There is project exists that I work for.
The project requires to write a scripts that fetch information from one Jira server and create an issue on other Jira server.
Current version of JiraPS does not allow such behaviors without call to Set-JiraConfigServer.
Therefore, the pull request allow to skip a switching Jira server between cmdlets calls.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [X] My code follows the code style of this project.
- [X] I have added Pester Tests that describe what my changes should do.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by timerplayer** [comment]
Found that all the failed test is result of running Help.Tests.ps1 file.
Of course, there are a things that I should changes in functions documentation.
I added a Session parameter in all public functions and did not change help docs.

But there are failed test exists that do no belong to my changes.
Therefore, I run same test on master branch and received near the same result.

What should I configure in order to pass all test for master branch?

File: JiraPS/Public/Invoke-JiraMethod.ps1 [hunk 1]
```diff
@@ -3,7 +3,7 @@ function Invoke-JiraMethod {
     [CmdletBinding( SupportsPaging )]
     param(
         [Parameter( Mandatory )]
-        [Uri]
+        [psobject]
```
Comment (lipkau) [inline JiraPS/Public/Invoke-JiraMethod.ps1]: `[Uri]` also supports relative paths. Why change the type of the variable?

Comment (timerplayer) [reply JiraPS/Public/Invoke-JiraMethod.ps1]: It is my background of .Net application development.

Uri class has a few [constructors](https://docs.microsoft.com/en-us/dotnet/api/system.uri.-ctor?view=netframework-4.8#System_Uri__ctor_System_String_), but any of them does not allow to create relative link without specify UriKind.
Therefore, the following examples will throw an exception: `[uri]::new("/test/a")` or ` New-Object uri "/test/a"`.

When I investigate all dependencies to Invoke-JiraMethod I found that often the module use string representation of URL, not an Uri instance.
So, in order to make our live easy and avoid the error I change the type to PSObject.

But, after I received you question I found that convert operation in PowerShell allow to convert relative path to Uri instance without specify UriKind: `[uri]"/test/a"`

In short: I will provide a new commit where I will change type back to URI.

**Comment by timerplayer** [comment]
> Please be aware that changing the behavior of public functions will be a breaking change (as you did by removing the -Credential parameter) and therefore require a new major release.
I supposed that it enough to add alias to -Session parameter.
Session parameter accept following types:
1. String - Name of a session
2. JiraPS.Session - a session itself
3. PSCredential - a credential that should be used for current operation.

>However, if a new major release is needed, I want the common session handling for all AtlassianPS modules to be used - which I already started to implement here
May you share a link where you describe an idea of common configuration I may assist you with the change for JIraPS.

>If you manage to add the functionality without changing current behavior, I will gladly release a new minor version with your changes
If alias is not enough to provide backward compatibility, I can return back -Credential parameter and use its value together with -Session parameter's value.
Will it suits for minor release?

**Comment by lipkau** [comment]
> May you share a link where you describe an idea of common configuration I may assist you with the change for JIraPS.
I did share the link here:
However, if a new major release is needed, I want the [common session handling for all AtlassianPS modules](https://github.com/AtlassianPS/AtlassianPS/issues/4) to be used - which I already started to implement [here](https://github.com/lipkau/ConfluencePS/tree/feature/useAtlassianPSConfiguration)

File: JiraPS/Public/Set-JiraIssueLabel.ps1 [hunk 2]
```diff
@@ -1,4 +1,4 @@
-﻿function Set-JiraIssueLabel {
+function Set-JiraIssueLabel {
```
Comment (lipkau) [inline JiraPS/Public/Set-JiraIssueLabel.ps1]: this change is switching from UTF8-BOM to UTF8, right?

interesting... as I will change all of the files to UTF8-BOM 😆

Comment (timerplayer) [reply JiraPS/Public/Set-JiraIssueLabel.ps1]: I started to use VS Code for PowerShell.
Therefore, I lose a control of encoding.

Just to confirm: Should I convert all files to UTF8-BOM?

Comment (lipkau) [reply JiraPS/Public/Set-JiraIssueLabel.ps1]: As you prefer. You can leave it as-is.

VS Code allows you to change the encoding of a file by clicking here:
image
or with "Change File Encoding" when pressing `Ctrl+Shift+P`

File: JiraPS/Public/Invoke-JiraMethod.ps1 [hunk 3]
```diff
@@ -73,7 +69,19 @@ function Invoke-JiraMethod {
         $_headers = Join-Hashtable -Hashtable $script:DefaultHeaders, $PSDefaultParameterValues["Invoke-WebRequest:Headers"], $Headers
         #endregion Headers
 
+        #region Session
+        # Convert a value provide by the argument to JiraPS.Session
+        $Session = ConvertTo-JiraSession -InputObject $Session
```
[the remark this answers is not in the document]
Comment (timerplayer) [reply JiraPS/Public/Invoke-JiraMethod.ps1]: It a bug.

Due to my plan ConvertTo-JiraSession should return default session if no session stored.
Therefore, the behavior for Invoke-JiraMethod have to be left the same as it was before my changes.

File: JiraPS/Public/Set-JiraIssueLabel.ps1 (hunk 2 above)
Comment (timerplayer) [reply JiraPS/Public/Set-JiraIssueLabel.ps1]: Thank you!

File: Tests/Functions/Add-JiraFilterPermission.Unit.Tests.ps1 [hunk 4]
```diff
@@ -50,18 +49,18 @@ Describe 'Add-JiraFilterPermission' -Tag 'Unit' {
     "id": 10010,
     "type": "project",
     "project": {
-      "self": "$jiraServer/jira/rest/api/2/project/EX",
+      "self": "jira/rest/api/2/project/EX",
```
[the remark this answers is not in the document]
Comment (timerplayer) [reply Tests/Functions/Add-JiraFilterPermission.Unit.Tests.ps1]: It was appear in bulk change when I removed absolute URL in mocks.
According to an idea Invoke-JiraMethod should receive relative URL for most common cases.

I seen the file on change list and left it as it is due to mock compares the same URL.
So, we ensure that Add-JiraFilterPermission bypass an URL from self field of response object.

To sum up: for cases like Add-JiraFilterPermission Unit Test we may reject part of my changes and restore absolute URL.

Should I process it?

[the remark this answers is not in the document]
Comment (lipkau) [reply Tests/Functions/Add-JiraFilterPermission.Unit.Tests.ps1]: yes. please do

**Comment by lipkau** [comment]
@timerplayer , any new on this?

**Comment by timerplayer** [comment]
Hello @lipkau,

Sorry, I had no time to provide a fixes, due to busy by other project.
What should I do?

**Comment by lipkau** [comment]
@timerplayer 
There were some major changes by atlassian to the api in order to be compliant to GDPR.
I have been investigating how to adapt this module to still work kind of the same for cloud servers and on-premise.
I changes had a lot of impact on the searching and fetching of users.
This is the current status (still WIP)
https://github.com/lipkau/JiraPS/blob/feature/ProperUserHandling/JiraPS/Public/Get-JiraUser.ps1
https://github.com/lipkau/JiraPS/blob/feature/ProperUserHandling/JiraPS/Public/Find-JiraUser.ps1

As I see, these changes also include what this PR would have changed.
If you agree, I would like to close this PR.
